src/irmin-pack/unix/io.ml

```diff
@@ -0,0 +1,282 @@
+                      (* third is parameters *)
+                      Printf.sprintf "path=%s; overwrite=%b" path overwrite )))
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let open_ ~path ~readonly : (t, [> open_error ]) result =
+    try
+      let ok = Sys.file_exists path in
+      assert ok;
+      let mode = Unix.(if readonly then O_RDONLY else O_RDWR) in
+      let fd = Unix.(openfile path [ mode; O_CLOEXEC ] default_open_perm) in
+      Ok { fd; closed = false; readonly; path }
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let close t : (unit, [> close_error ]) result =
+    try
+      match t.closed with
+      | true -> Error `Double_close
+      | false ->
+          t.closed <- true;
+          Unix.close t.fd;
+          Ok ()
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let write_string t ~off s : (unit, [> write_error ]) result =
+    try
+      match (t.closed, t.readonly) with
+      | true, _ -> Error `Write_on_closed
+      | _, true -> Error `Ro_not_allowed
+      | _ ->
+          (* really_write and following do not mutate the given buffer, so
+             Bytes.unsafe_of_string is actually safe *)
+          let buf = Bytes.unsafe_of_string s in
+          let () = Util.really_write t.fd off buf 0 (Bytes.length buf) in
+          Ok ()
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  (* NOTE copied from write_string above; ensure the code for both functions remains in
+     sync *)
+  let write_exn t ~off s : unit =
+    try
+      match (t.closed, t.readonly) with
+      | true, _ -> raise (Write_error `Write_on_closed)
+      | _, true -> raise (Write_error `Ro_not_allowed)
+      | _ ->
+          (* really_write and following do not mutate the given buffer, so
+             Bytes.unsafe_of_string is actually safe *)
+          let buf = Bytes.unsafe_of_string s in
+          let () = Util.really_write t.fd off buf 0 (Bytes.length buf) in
+          ()
+    with Unix.Unix_error (e, s1, s2) ->
+      raise (Write_error (`Io_misc (e, s1, s2)))
+
+  let fsync t : (unit, [> write_error ]) result =
+    try
+      match (t.closed, t.readonly) with
+      | true, _ -> Error `Write_on_closed
+      | _, true -> Error `Ro_not_allowed
+      | _ ->
+          Unix.fsync t.fd;
+          Ok ()
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let read_to_string t ~off ~len =
+    try
+      match t.closed with
+      | true -> Error `Read_on_closed
+      | false -> (
+          let buf = Bytes.create len in
+          let nread = Util.really_read t.fd off len buf in
+          match nread = len with
+          | false ->
+              (* didn't manage to read the desired amount; in this case the interface seems to
+                 require we return `Read_out_of_bounds FIXME check this, because it is unusual
+                 - the normal API allows return of a short string *)
```

Comment:
@tomjridge I think that the current behaviour is more predictable. But if we ever need the feature where reading less is not an error, we should change it back. 

Do you think this is OK to leave it as is?

Reply:
Yes, sure! I just flagged it as unusual because I wasn't completely sure what you wanted.